pbft/pbft.cpp

```diff
@@ -978,6 +992,533 @@ pbft::get_high_water_mark()
+            {
+                LOG(error) << "is_valid_viewchange_message - a pre prepare message has invalid sequence, view or request hash";
+                return false;
+            }
+
+            senders.insert(prepare_envelope.sender());
+        }
+
+        if (senders.size() < 2 * this->max_faulty_nodes() + 1)
+        {
+            LOG(error) << "is_valid_viewchange_message - not enough prepares in a prepared_proof";
+            return false;
+        }
+    }
+    return true;
+}
+
+bool
+pbft::get_sequences_and_request_hashes_from_proofs(
+        const pbft_msg& viewchange_msg
+        , std::set<std::pair<uint64_t, std::string>>& sequence_request_pairs) const
+{
+    for (int j{0}; j < viewchange_msg.prepared_proofs_size(); ++j)
+    {
+        // --- add the sequence/hash to a set
+        auto prepared_proof = viewchange_msg.prepared_proofs(j);
+
+        pbft_msg msg;
+        if (bzn_envelope envelope{prepared_proof.pre_prepare()}; /*!envelope.ParseFromString(prepared_proof.pre_prepare()) ||*/ !this->is_peer(envelope.sender()) || !msg.ParseFromString(envelope.pbft()))
+        {
+            return false;
+        }
+        sequence_request_pairs.insert(std::make_pair(msg.sequence(), msg.request_hash()));
+    }
+    return true;
+}
+
+/**
+ * A valid newview message must contain a complete set of pre prepares, that is
+ * there must be one pre prepare for each sequence.
+ * TODO: this method needs a better name
+ * @param newview_msg
+ * @return true if there are no missing pre-prepares based on sequence numbers of each pre-prepare, false otherwise.
+ */
+bool
+pbft::pre_prepares_contiguous(uint64_t latest_sequence, const pbft_msg& newview_msg)
+{
+    if (0 == newview_msg.pre_prepare_messages_size())
+    {
+        return true;
+    }
+
+    auto extract_sequence = [](const bzn_envelope pre_prepare_envelope)->uint64_t
+            {
+                pbft_msg pre_prepare_message;
+                pre_prepare_message.ParseFromString(pre_prepare_envelope.pbft());
+                return pre_prepare_message.sequence();
+            };
+
+    // We are assuming that the pre prepare messages are stored sequentially by sequence number in newview_msg.pre_prepare_messages
+    uint64_t last_sequence{extract_sequence(newview_msg.pre_prepare_messages(0))};
+    for (int i{1}; i < newview_msg.pre_prepare_messages_size(); ++i)
+    {
+        uint64_t current_sequence{extract_sequence(newview_msg.pre_prepare_messages(i))};
+        if (last_sequence + 1 != current_sequence)
+        {
+            return false;
+        }
+        last_sequence = current_sequence;
+    }
+    return last_sequence==latest_sequence;
+}
+
+// prepared_proofs is not a member of a new_view message. You need to
+// iterate over all the prepared_proofs in all the viewchange messages
+// in the new_view.
+uint64_t
+pbft::last_sequence_in_newview_preprepare_messages(const pbft_msg &newview)
```

Comment:
another casualty, removing.

Reply:
why keep the code in?  Also, what about the check for contiguous pre-prepares?

Reply:
Whups, removed the commented code.

Isabel and I rewrote the code that used that function to make it simpler. It may be that we neglected to do the test incorrectly.